src/software/simulation/simulator.cpp

```diff
@@ -11,63 +11,88 @@ extern "C"
-        yellow_robots.emplace_back(state);
-    }
-    physics_world.addYellowRobots(yellow_robots);
-    std::vector<RobotStateWithId> blue_robots;
-    for (const auto& robot : world.enemyTeam().getAllRobots())
+    physics_world.setBallState(ball_state);
+    simulator_ball = std::make_shared<SimulatorBall>(physics_world.getPhysicsBall());
+}
+
+void Simulator::removeBall()
+{
+    simulator_ball.reset();
+    physics_world.removeBall();
+}
+
+void Simulator::addYellowRobots(const std::vector<RobotStateWithId>& robots)
+{
+    physics_world.addYellowRobots(robots);
+    for (auto physics_robot : physics_world.getYellowPhysicsRobots())
     {
-        RobotStateWithId state{.id          = robot.id(),
-                               .robot_state = robot.currentState().robotState()};
-        blue_robots.emplace_back(state);
+        auto simulator_robot = std::make_shared<SimulatorRobot>(physics_robot);
+        auto firmware_robot  = SimulatorRobotSingleton::createFirmwareRobot();
+        auto firmware_ball   = SimulatorBallSingleton::createFirmwareBall();
+        // Release ownership of the pointers so the firmware_world can take ownership
+        FirmwareWorld_t* firmware_world_raw =
+            app_firmware_world_create(firmware_robot.release(), firmware_ball.release());
+        auto firmware_world =
+            std::shared_ptr<FirmwareWorld_t>(firmware_world_raw, FirmwareWorldDeleter());
+
+        yellow_simulator_robots.insert(std::make_pair(simulator_robot, firmware_world));
     }
-    physics_world.addBlueRobots(blue_robots);
+}
 
-    for (auto physics_robot : physics_world.getFriendlyPhysicsRobots())
+void Simulator::addBlueRobots(const std::vector<RobotStateWithId>& robots)
+{
+    physics_world.addBlueRobots(robots);
+    for (auto physics_robot : physics_world.getBluePhysicsRobots())
     {
         auto simulator_robot = std::make_shared<SimulatorRobot>(physics_robot);
-
-        auto firmware_robot = SimulatorRobotSingleton::createFirmwareRobot();
-        auto firmware_ball  = SimulatorBallSingleton::createFirmwareBall();
+        auto firmware_robot  = SimulatorRobotSingleton::createFirmwareRobot();
+        auto firmware_ball   = SimulatorBallSingleton::createFirmwareBall();
         // Release ownership of the pointers so the firmware_world can take ownership
         FirmwareWorld_t* firmware_world_raw =
             app_firmware_world_create(firmware_robot.release(), firmware_ball.release());
         auto firmware_world =
             std::shared_ptr<FirmwareWorld_t>(firmware_world_raw, FirmwareWorldDeleter());
 
-        simulator_robots.insert(std::make_pair(simulator_robot, firmware_world));
+        blue_simulator_robots.insert(std::make_pair(simulator_robot, firmware_world));
     }
-
-    simulator_ball = std::make_shared<SimulatorBall>(physics_world.getPhysicsBall());
 }
 
-void Simulator::stepSimulation(const Duration& time_step)
+void Simulator::setYellowRobotPrimitives(ConstPrimitiveVectorPtr primitives)
```

Comment:
this is duplicated w/ BlueRobotPrimitives exactly other than the `yellow_simulator_robots` and `blue_simulator_robots`. Maybe its worth having a helper, and pass In the vector as an arg?

Reply:
Done, and did the same thing for `addRobots`. I did need to set variables by reference so please take a look and lmk if you're ok with that.